### PR TITLE
Fixed the problem of Odo not ignoring files that it can't find

### LINKED ISSUE #966: odo watch should ignore files it can't find
[kind/bug]

## What versions of software are you using?

0.0.14

Operating System :

CentOS

Output of `odo version` : 

```
$ odo version
odo v0.0.14 (618c6d3)

Server: https://172.30.0.1:443
OpenShift: v3.11.16
Kubernetes: v1.11.0+d4cacc0
```

## How did you run odo exactly?

Am using watch behaviour by running:

```
odo watch
```

The problem is that if a temporary file is created, which is then deleted before odo processes the file, odo will raise an error and exit rather than ignore the file.

```
$ odo watch
Waiting for something to change in /opt/app-root/src/frontend
E1113 05:34:30.462573     612 watch.go:147] Failed getting details of the changed file /opt/app-root/src/frontend/sedDab8B9. Ignoring the change
E1113 05:34:30.462701     612 watch.go:153] Ignoring event for file /opt/app-root/src/frontend/sedDab8B9 as details about the file couldn't be fetched
Error while trying to watch /opt/app-root/src/frontend
$
```

This can occur when run a command like:

```
sed -i "s/Wild West/Grumpy West/" index.html
```

This results in sed creating a temporary file with edits before renaming it to the original. The ``sedDab8B9`` file it complains about is this temporary file. Because it only existed for a moment, odo couldn't find it and exited instead of ignoring it.

## Expected behavior

Should ignore a file it was notified about but which was removed before processing.

## Actual behavior

Exits the ``odo watch`` command.

## Steps to reproduce the behavior

Run:

```
odo watch
```

In another window run sed command such as:

```
sed -i "s/Wild West/Grumpy West/" index.html
```

in file in the directory.

## Any logs, error output, etc?

See above.

## Do you have any idea what may have caused this?

Not ignoring files which it finds removed after notified of them.

## Do you have an idea how to solve the issue?

Ignore files which it can't find after being notified of them.

### PR DESCRIPTION
**Pull request by mik-dass**
fixes #966

### CODE REVIEW

**Comment by geoand** [comment]
@mik-dass Would it be possible to add a test case for this?

**Comment by mik-dass** [comment]
@geoand Not sure, I will try though :)

**Comment by geoand** [comment]
Thanks @mik-dass !

File: pkg/component/watch.go [hunk 1]
```diff
@@ -73,9 +73,17 @@ func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string)
 		}
 		if !matched {
 			glog.V(4).Infof("adding watch on path %s", path)
+
+			// checking if the file exits before adding the watcher to it
+			if _, err := os.Stat(path); os.IsNotExist(err) {
```
Comment (metacosm) [inline pkg/component/watch.go]: Extract this into a function?

File: pkg/component/watch.go [hunk 2]
```diff
@@ -73,9 +73,17 @@ func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string)
...
+				// path to file does not exist
+				glog.V(4).Infof("path %s doesn't exists, skipping it", path)
```
Comment (metacosm) [inline pkg/component/watch.go]: Typo: s/exists/exist/

File: pkg/component/watch.go [hunk 3]
```diff
@@ -114,14 +122,22 @@ func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string)
 			glog.V(4).Infof("ignoring watch for %s", v)
 			continue
 		}
+
+		// checking if the file exits before adding the watcher to it
+		if _, err := os.Stat(path); os.IsNotExist(err) {
+			// path to file does not exist
+			glog.V(4).Infof("path %s doesn't exists currently, skipping it", path)
```
Comment (metacosm) [inline pkg/component/watch.go]: Typo as above.

**Comment by mik-dass** [comment]
@metacosm Fixed. Please have a look again :)

File: pkg/component/watch.go [hunk 4]
```diff
@@ -75,10 +76,8 @@ func addRecursiveWatch(watcher *fsnotify.Watcher, path string, ignores []string)
 			glog.V(4).Infof("adding watch on path %s", path)
 
 			// checking if the file exits before adding the watcher to it
-			if _, err := os.Stat(path); os.IsNotExist(err) {
-				// path to file does not exist
-				glog.V(4).Infof("path %s doesn't exists, skipping it", path)
-				return nil
+			if !util.CheckPathExists(path) {
+				glog.V(4).Infof("path %s doesn't exist, skipping it", path)
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/component/watch.go]: Fixed

**Comment by geoand** [comment]
Let's merge this when the tests pass :)

**Pull request merged**